### ISSUE TITLE
NDI Dedicated output on Scene does not include Audio

### ISSUE THREAD

**Issue by ramlax13** [obs-ndi/obs-ndi#204]
Is it possible to use NDI Dedicated Output as a Scene filter and have the audio for the scene go along with it?  I'm trying to output Video and Audio as a single NDI signal from OBS, but that also needs to include Alpha.  Is this at all possible?

**Comment by Helgaiden** [comment]
Add a dedicated audio only NDI filter to the audio source and add both sources on the streaming PC/receiving PC. Maybe that'll work well enough.

**Comment by Trouffman** [comment]
Having this challenge as well.
Using the audio source out can cause sync problems (depending on how your receiver is getting the feed.

I see this been flagged in oct 2018. not sure if there is a fix since or a duplicate of this issue.

**Comment by paulpv** [comment]
This seems fixed in https://github.com/Palakis/obs-ndi/pull/849 's 4.11.0-RC; Closing.
Feel free to comment if you test and disagree and I should reopen.

**Comment by FirstzeroOne** [comment]
Still Not working or not working again. Tested with
OBS Versions 29.0.1 29.0.2 29.1rc1 
NDI Tools Version 5.5.4.0 and Runtime 5.5.3
Plugin Version 4.11.1 4.11.0 4.11.0rc 

Audio from Source is always working, but never from scene.

**Comment by paulpv** [comment]
`Today I Learned` that you can add a Filter to a Scene!
I guess I looked into this before, but today I do not remember this being a feature of OBS.
I can look into it again.

**Comment by Pzx5073** [comment]
I tried scene source filter and branch-output filter a while ago and they can stream video and audio from the scene filter, could that be implemented for ndi plugin at some point? It would be a very useful feature.

**Comment by Trouffman** [comment]
To all following this issue, please report if this is still the case on latest DistroAV release

**Comment by piaocn** [comment]
> To all following this issue, please report if this is still the case on latest DistroAV release

Hello Trouffman, I try a lot on OBS 31.0.0 with DistroAV 6.0.0, it is still not possible to send NDI video along with audio on scene on OBS, while the main Output of DistroAV NDI setting works fine with both audio and video. I think this bug is still needed to be sovled.

**Comment by piaocn** [comment]
> To all following this issue, please report if this is still the case on latest DistroAV release

Here is my log file:

20:51:24.024: CPU Name: AMD Ryzen 7 5800X 8-Core Processor             
20:51:24.024: CPU Speed: 3800MHz
20:51:24.024: Physical Cores: 8, Logical Cores: 16
20:51:24.024: Physical Memory: 32694MB Total, 25007MB Free
20:51:24.024: Windows Version: 10.0 Build 19045 (release: 22H2; revision: 5131; 64-bit)
20:51:24.024: Running as administrator: false
20:51:24.024: Windows 10/11 Gaming Features:
20:51:24.024: 	Game DVR: On
20:51:24.024: 	Game Mode: Probably On (no reg key set)
20:51:24.025: Sec. Software Status:
20:51:24.026: 	火绒安全软件: enabled (AV)
20:51:24.026: 	Microsoft Defender Antivirus: disabled (AV)
20:51:24.026: 	Windows Firewall: enabled (FW)
20:51:24.027: Current Date/Time: 2024-12-12, 20:51:24
20:51:24.027: Browser Hardware Acceleration: true
20:51:24.027: Hide OBS windows from screen capture: false
20:51:24.027: Qt Version: 6.6.3 (runtime), 6.6.3 (compiled)
20:51:24.027: Portable mode: false
20:51:24.225: OBS 31.0.0 (64-bit, windows)
20:51:24.225: ---------------------------------
20:51:24.284: ---------------------------------
20:51:24.284: audio settings reset:
20:51:24.284: 	samples per sec: 48000
20:51:24.284: 	speakers:        2
20:51:24.284: 	max buffering:   960 milliseconds
20:51:24.284: 	buffering type:  dynamically increasing
20:51:24.285: ---------------------------------
20:51:24.285: Initializing D3D11...
20:51:24.285: Available Video Adapters: 
20:51:24.287: 	Adapter 0: NVIDIA GeForce RTX 3080 Ti
20:51:24.287: 	  Dedicated VRAM: 12673089536 (11.8 GiB)
20:51:24.287: 	  Shared VRAM:    17141121024 (16.0 GiB)
20:51:24.287: 	  PCI ID:         10de:2208
20:51:24.287: 	  HAGS Status:    Disabled (Default: No, Driver status: Supported)
20:51:24.287: 	  Driver Version: 32.0.15.6094
20:51:24.288: 	  output 0:
20:51:24.288: 	    name=CG321QKS
20:51:24.288: 	    pos={0, 0}
20:51:24.288: 	    size={2560, 1440}
20:51:24.288: 	    attached=true
20:51:24.288: 	    refresh=120
20:51:24.288: 	    bits_per_color=8
20:51:24.288: 	    space=RGB_FULL_G22_NONE_P709
20:51:24.288: 	    primaries=[r=(0.673828, 0.317383), g=(0.280273, 0.660156), b=(0.142578, 0.044922), wp=(0.312500, 0.329102)]
20:51:24.288: 	    relative_gamut_area=[709=1.291061, P3=0.951732, 2020=0.682804]
20:51:24.288: 	    sdr_white_nits=80
20:51:24.288: 	    nit_range=[min=0.069800, max=417.711792, max_full_frame=391.463287]
20:51:24.288: 	    dpi=120 (125%)
20:51:24.288: 	    id=\\?\DISPLAY#HKC3234#5&3b7692c&0&UID176384#{e6f07b5f-ee97-4a90-b076-33f57bf4eaa7}
20:51:24.288: 	    alt_id=\\.\DISPLAY1
20:51:24.288: 	Adapter 1: NVIDIA GeForce RTX 3080 Ti
20:51:24.288: 	  Dedicated VRAM: 12673089536 (11.8 GiB)
20:51:24.288: 	  Shared VRAM:    17141121024 (16.0 GiB)
20:51:24.288: 	  PCI ID:         10de:2208
20:51:24.288: 	  HAGS Status:    Disabled (Default: No, Driver status: Supported)
20:51:24.288: 	  Driver Version: 32.0.15.6094
20:51:24.288: Loading up D3D11 on adapter NVIDIA GeForce RTX 3080 Ti (0)
20:51:24.385: D3D11 loaded successfully, feature level used: b000
20:51:24.385: DXGI increase maximum frame latency success
20:51:24.385: D3D11 GPU priority setup failed (not admin?)
20:51:24.457: ---------------------------------
20:51:24.457: video settings reset:
20:51:24.457: 	base resolution:   1920x1080
20:51:24.457: 	output resolution: 1920x1080
20:51:24.457: 	downscale filter:  Bicubic
20:51:24.457: 	fps:               30/1
20:51:24.457: 	format:            NV12
20:51:24.457: 	YUV mode:          Rec. 709/Partial
20:51:24.458: NV12 texture support enabled
20:51:24.458: P010 texture support not available
20:51:24.459: Audio monitoring device:
20:51:24.459: 	name: Default
20:51:24.459: 	id: default
20:51:24.460: ---------------------------------
20:51:24.461: No AJA devices found, skipping loading AJA UI plugin
20:51:24.461: Failed to initialize module 'aja-output-ui.dll'
20:51:24.462: No AJA devices found, skipping loading AJA plugin
20:51:24.462: Failed to initialize module 'aja.dll'
20:51:24.463: [Audio Monitor] loaded version 0.8.5
20:51:24.475: Skipping module '../../obs-plugins/64bit/chrome_elf.dll', not an OBS plugin
20:51:24.477: [CoreAudio encoder]: CoreAudio AAC encoder not installed on the system or couldn't be loaded
20:51:24.478: Failed to load 'en-US' text for module: 'decklink-captions.dll'
20:51:24.478: Failed to load 'en-US' text for module: 'decklink-output-ui.dll'
20:51:24.480: A DeckLink iterator could not be created.  The DeckLink drivers may not be installed
20:51:24.480: Failed to initialize module 'decklink.dll'
20:51:24.481: [DistroAV] obs_module_load: you can haz DistroAV (Version 6.0.0)
20:51:24.481: [DistroAV] obs_module_load: Qt Version: 6.6.3 (runtime), 6.6.2 (compiled)
20:51:24.481: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:24.481: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:24.482: [DistroAV] load_ndilib: Trying 'C:\Program Files\NDI\NDI 6 Tools\Runtime\Processing.NDI.Lib.x64.dll'
20:51:24.482: [DistroAV] load_ndilib: Found 'C:\Program Files\NDI\NDI 6 Tools\Runtime\Processing.NDI.Lib.x64.dll'; attempting to load NDI library...
20:51:24.484: [DistroAV] load_ndilib: NDI library loaded successfully
20:51:24.484: [DistroAV] load_ndilib: NDIlib_v5_load found
20:51:24.492: [DistroAV] obs_module_load: NDI library initialized successfully ('NDI SDK WIN64 12:30:38 Dec  4 2024 6.1.0.0')
20:51:24.526: Skipping module '../../obs-plugins/64bit/libcef.dll', not an OBS plugin
20:51:24.526: Skipping module '../../obs-plugins/64bit/libEGL.dll', not an OBS plugin
20:51:24.526: Skipping module '../../obs-plugins/64bit/libGLESv2.dll', not an OBS plugin
20:51:24.527: [NVIDIA Audio Effects:] NVIDIA denoiser disabled, redistributable not found or could not be loaded.
20:51:24.527: Failed to get NVVideoEffects.dll version info size
20:51:24.528: [NVIDIA VIDEO FX]: FX disabled, redistributable not found or could not be loaded.
20:51:24.533: [obs-browser]: Version 2.24.4
20:51:24.533: [obs-browser]: CEF Version 127.0.6533.100 (runtime), 127.3.4+ge9e2e14+chromium-127.0.6533.100 (compiled)
20:51:24.769: [obs-nvenc] NVENC version: 12.2 (compiled) / 12.2 (driver), CUDA driver version: 12.60, AV1 supported: false
20:51:24.805: [obs-websocket] [obs_module_load] you can haz websockets (Version: 5.5.4 | RPC Version: 1)
20:51:24.805: [obs-websocket] [obs_module_load] Qt version (compile-time): 6.6.3 | Qt version (run-time): 6.6.3
20:51:24.805: [obs-websocket] [obs_module_load] Linked ASIO Version: 103100
20:51:24.805: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:24.813: [obs-websocket] [obs_module_load] Module loaded.
20:51:24.816: [vlc-video]: Couldn't find VLC installation, VLC video source disabled
20:51:24.823: ---------------------------------
20:51:24.823:   Loaded Modules:
20:51:24.823:     win-wasapi.dll
20:51:24.823:     win-dshow.dll
20:51:24.823:     win-capture.dll
20:51:24.823:     vlc-video.dll
20:51:24.823:     text-freetype2.dll
20:51:24.823:     rtmp-services.dll
20:51:24.823:     obs-x264.dll
20:51:24.823:     obs-websocket.dll
20:51:24.823:     obs-webrtc.dll
20:51:24.823:     obs-vst.dll
20:51:24.823:     obs-transitions.dll
20:51:24.823:     obs-text.dll
20:51:24.823:     obs-qsv11.dll
20:51:24.823:     obs-outputs.dll
20:51:24.823:     obs-nvenc.dll
20:51:24.823:     obs-filters.dll
20:51:24.823:     obs-ffmpeg.dll
20:51:24.823:     obs-browser.dll
20:51:24.823:     nv-filters.dll
20:51:24.823:     image-source.dll
20:51:24.823:     frontend-tools.dll
20:51:24.823:     distroav.dll
20:51:24.823:     decklink-output-ui.dll
20:51:24.823:     decklink-captions.dll
20:51:24.823:     coreaudio-encoder.dll
20:51:24.823:     audio-monitor.dll
20:51:24.823: ---------------------------------
20:51:24.823: [DistroAV] +obs_module_post_load()
20:51:24.823: [DistroAV] +updateCheckStart(userRequestCallback=nullptr)
20:51:24.823: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:24.823: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:24.823: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:24.823: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:24.823: [DistroAV] updateCheckStart: elapsedSeconds=6768 < minAutoUpdateCheckIntervalSeconds=86400; ignoring
20:51:24.823: [DistroAV] -updateCheckStart(userRequestCallback=nullptr)
20:51:24.823: [DistroAV] -obs_module_post_load()
20:51:24.823: ---------------------------------
20:51:24.823: Available Encoders:
20:51:24.823:   Video Encoders:
20:51:24.823: 	- ffmpeg_svt_av1 (SVT-AV1)
20:51:24.823: 	- ffmpeg_aom_av1 (AOM AV1)
20:51:24.823: 	- obs_nvenc_h264_tex (NVIDIA NVENC H.264)
20:51:24.823: 	- obs_nvenc_hevc_tex (NVIDIA NVENC HEVC)
20:51:24.823: 	- obs_x264 (x264)
20:51:24.823:   Audio Encoders:
20:51:24.823: 	- ffmpeg_aac (FFmpeg AAC)
20:51:24.823: 	- ffmpeg_opus (FFmpeg Opus)
20:51:24.823: 	- ffmpeg_pcm_s16le (FFmpeg PCM (16-bit))
20:51:24.823: 	- ffmpeg_pcm_s24le (FFmpeg PCM (24-bit))
20:51:24.823: 	- ffmpeg_pcm_f32le (FFmpeg PCM (32-bit float))
20:51:24.823: 	- ffmpeg_alac (FFmpeg ALAC (24-bit))
20:51:24.823: 	- ffmpeg_flac (FFmpeg FLAC (16-bit))
20:51:24.823: ==== Startup complete ===============================================
20:51:24.904: All scene data cleared
20:51:24.904: ------------------------------------------------
20:51:24.907: [DistroAV] +ndi_filter_getdefaults(...)
20:51:24.907: [DistroAV] -ndi_filter_getdefaults(...)
20:51:24.907: [DistroAV] +ndi_filter_create(name=`s2`, groups=``)
20:51:24.907: [DistroAV] +ndi_filter_update(name=`scene 2`)
20:51:24.912: [DistroAV] -ndi_filter_update(name=`scene 2`)
20:51:24.912: [DistroAV] -ndi_filter_create(...)
20:51:24.912: [DistroAV] +ndi_filter_getdefaults(...)
20:51:24.912: [DistroAV] -ndi_filter_getdefaults(...)
20:51:24.912: [win-wasapi: 'Audio Output Capture'] update settings:
20:51:24.912: 	device id: {0.0.0.00000000}.{e7dbce47-33cf-448c-915b-6a8409d48985}
20:51:24.912: 	use device timing: 1
20:51:24.913: [Media Source 'M31-1021.mov']: settings:
20:51:24.913: 	input:                   C:/Users/RythenBox/Desktop/2024.12.03/M31-1021.mov
20:51:24.913: 	input_format:            (null)
20:51:24.913: 	speed:                   100
20:51:24.913: 	is_looping:              no
20:51:24.913: 	is_linear_alpha:         no
20:51:24.913: 	is_hw_decoding:          no
20:51:24.913: 	is_clear_on_media_end:   yes
20:51:24.913: 	restart_on_activate:     yes
20:51:24.913: 	close_when_inactive:     no
20:51:24.913: 	full_decode:             no
20:51:24.913: 	ffmpeg_options:          
20:51:24.913: [Media Source '01.我和我的祖国.flac']: settings:
20:51:24.913: 	input:                   C:/Users/RythenBox/Desktop/01.我和我的祖国.flac
20:51:24.913: 	input_format:            (null)
20:51:24.913: 	speed:                   100
20:51:24.913: 	is_looping:              no
20:51:24.913: 	is_linear_alpha:         no
20:51:24.913: 	is_hw_decoding:          no
20:51:24.913: 	is_clear_on_media_end:   yes
20:51:24.913: 	restart_on_activate:     yes
20:51:24.913: 	close_when_inactive:     no
20:51:24.913: 	full_decode:             no
20:51:24.913: 	ffmpeg_options:          
20:51:24.913: [DistroAV] +ndi_filter_getdefaults(...)
20:51:24.913: [DistroAV] -ndi_filter_getdefaults(...)
20:51:24.913: [DistroAV] +ndi_filter_create_audioonly(name=`Dedicated NDI® Output`, groups=``)
20:51:24.913: [DistroAV] +ndi_filter_update(name=`Dedicated NDI® output (Audio Only)`)
20:51:24.915: [DistroAV] -ndi_filter_update(name=`Dedicated NDI® output (Audio Only)`)
20:51:24.915: [DistroAV] -ndi_filter_create_audioonly(...)
20:51:24.915: [DistroAV] +ndi_filter_getdefaults(...)
20:51:24.915: [DistroAV] -ndi_filter_getdefaults(...)
20:51:24.918: Switched to scene 'Scene 2'
20:51:24.918: ------------------------------------------------
20:51:24.918: Loaded scenes:
20:51:24.918: - scene 'Scene 2':
20:51:24.918:     - source: 'Audio Output Capture' (wasapi_output_capture)
20:51:24.918:     - source: '01.我和我的祖国.flac' (ffmpeg_source)
20:51:24.918:         - filter: 'Dedicated NDI® output (Audio Only)' (ndi_audiofilter)
20:51:24.918:     - source: 'Color Source' (color_source_v3)
20:51:24.918:     - source: 'M31-1021.mov' (ffmpeg_source)
20:51:24.918:     - filter: 'scene 2' (ndi_filter)
20:51:24.918: - scene 'Scene 3':
20:51:24.918: ------------------------------------------------
20:51:24.930: WASAPI: Device '耳机 (Mi TV soundbar Stereo)' [44100 Hz] initialized (source: Audio Output Capture)
20:51:24.961: [DistroAV] ndi_filter_asyncaudio: growing audio_conv_buffer from 0 to 35528 bytes
20:51:24.961: [DistroAV] ndi_filter_asyncaudio: allocating 35528 bytes
20:51:25.052: [DistroAV] ndi_filter_asyncaudio: growing audio_conv_buffer from 35528 to 35672 bytes
20:51:25.052: [DistroAV] ndi_filter_asyncaudio: freeing 35528 bytes
20:51:25.052: [DistroAV] ndi_filter_asyncaudio: allocating 35672 bytes
20:51:25.272: [DistroAV] +main_output_init()
20:51:25.272: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:25.272: [DistroAV] +main_output_deinit()
20:51:25.272: [DistroAV] main_output_deinit: NDI main output is not initialized
20:51:25.272: [DistroAV] -main_output_deinit()
20:51:25.272: [DistroAV] main_output_init: creating NDI main output 'OBS'
20:51:25.272: [DistroAV] +ndi_output_getdefaults()
20:51:25.272: [DistroAV] -ndi_output_getdefaults()
20:51:25.272: [DistroAV] +ndi_output_create(name='OBS', groups='', ...)
20:51:25.272: [DistroAV] ndi_output_update(name='OBS', groups='', ...)
20:51:25.272: [DistroAV] -ndi_output_create(name='OBS', groups='', ...)
20:51:25.272: [DistroAV] main_output_init: successfully created NDI main output 'OBS'
20:51:25.272: [DistroAV] -main_output_init()
20:51:25.272: [DistroAV] +preview_output_init()
20:51:25.272: DEPRECATION: obs_frontend_get_global_config is deprecated. Read from global or user configuration explicitly instead.
20:51:25.272: [DistroAV] +preview_output_deinit()
20:51:25.272: [DistroAV] preview_output_deinit: NDI preview output is not initialized
20:51:25.272: [DistroAV] -preview_output_deinit()
20:51:25.272: [DistroAV] preview_output_init: creating NDI preview output 'OBS Preview'
20:51:25.272: [DistroAV] +ndi_output_getdefaults()
20:51:25.272: [DistroAV] -ndi_output_getdefaults()
20:51:25.272: [DistroAV] +ndi_output_create(name='OBS Preview', groups='', ...)
20:51:25.272: [DistroAV] ndi_output_update(name='OBS Preview', groups='', ...)
20:51:25.272: [DistroAV] -ndi_output_create(name='OBS Preview', groups='', ...)
20:51:25.272: [DistroAV] preview_output_init: successfully created NDI preview output 'OBS Preview'
20:51:25.272: [DistroAV] -preview_output_init()

**Comment by BitRate27** [comment]
TL;DR This is not supported by OBS

I am able to duplicate this issue. Our filter_audio callback is not being called when the filter is on a scene. I added OBS_SOURCE_AUDIO to the output flags, and I am not able to select the filter for a scene. This leads me to the conclusion that OBS does not support audio filters on scenes. I am curious why our filter_audio callback is called when our output_flags only specify video.

**Comment by BitRate27** [comment]
Further investigation has showed that other plugins can do this but they use backend, undocumented OBS APIs. One option is to put an audio mixing filter before the Dedicated NDI Output filter on the scene.

**Comment by MarximusMaximus** [comment]
@BitRate27 
> One option is to put an audio mixing filter before the Dedicated NDI Output filter on the scene.

Apologies, when I go to add filters for the scene, I don't see something like this. Could you post a screenshot of what you mean? (Do I need an additional plugin?) Thank you very much in advance.

macOS, OBS 31, DistroAV 6

**Comment by BitRate27** [comment]
It was only a thought, never tried it and cannot even say it will work but the https://obsproject.com/forum/resources/branch-output-streaming-recording-filter-for-source-scene.1987/ might work before the NDI Output Filter.

**Comment by MarximusMaximus** [comment]
Thank you, I'll give it a shot!

**Comment by lifuqiang0118** [comment]
Yes, I have encountered the same issue. I need to run a livestream within a virtual machine, but since the virtual machine can't access the GPU, my CPU becomes overloaded. I want to run OBS on the host machine and use a plugin to forward multiple scenes to different virtual machines. However, I'm now facing this issue which prevents me from continuing my livestreams. It seems my CPU might remain under heavy load indefinitely. I hope this can be optimized in a future update. It might be challenging, but I would be truly grateful if this could be improved in an upcoming release.

**Comment by RayneYoruka** [comment]
> Yes, I have encountered the same issue. I need to run a livestream within a virtual machine, but since the virtual machine can't access the GPU, my CPU becomes overloaded. I want to run OBS on the host machine and use a plugin to forward multiple scenes to different virtual machines. However, I'm now facing this issue which prevents me from continuing my livestreams. It seems my CPU might remain under heavy load indefinitely. I hope this can be optimized in a future update. It might be challenging, but I would be truly grateful if this could be improved in an upcoming release.

Thats... OBS on it's own. You need a GPU to process the scene. Regardless. You should see of GPu passthrough. 
**Completely unrelated to the thread.**
